presto-main/src/main/java/com/facebook/presto/operator/aggregation/fixedhistogram/FixedDoubleBreakdownHistogram.java

```diff
@@ -0,0 +1,242 @@
+
+import java.util.Iterator;
+import java.util.NoSuchElementException;
+
+import static com.google.common.base.Preconditions.checkArgument;
+
+public class FixedDoubleBreakdownHistogram
+        implements Cloneable, Iterable<FixedDoubleBreakdownHistogram.Bucket>
+{
+    private int bucketCount;
+    private double min;
+    private double max;
+    double tolerance;
+    private int size;
+
+    protected int[] indices;
+    protected double[] weights;
+    protected long[] counts;
+
+    public static class Bucket
+    {
+        public final double left;
+        public final double right;
+        public final double weight;
+        public final long count;
+
+        public Bucket(double left, double right, double weight, long count)
+        {
+            this.left = left;
+            this.right = right;
+            this.weight = weight;
+            this.count = count;
+        }
+    }
+
+    public FixedDoubleBreakdownHistogram(int bucketCount, double min, double max, double tolerance)
+    {
+        this.bucketCount = bucketCount;
+        this.min = min;
+        this.max = max;
+        Utils.verifyParams(bucketCount, min, max);
+        this.tolerance = tolerance;
+        size = 0;
+        verifyParams();
+        this.indices = new int[0];
+        this.weights = new double[0];
+        this.counts = new long[0];
+    }
+
+    public FixedDoubleBreakdownHistogram(SliceInput input)
+    {
+        this.bucketCount = input.readInt();
+        this.min = input.readDouble();
+        this.max = input.readDouble();
+        Utils.verifyParams(bucketCount, min, max);
+        this.tolerance = input.readDouble();
+        size = input.readInt();
+        verifyParams();
+        this.weights = new double[size];
+        this.counts = new long[size];
+        input.readBytes(
+                Slices.wrappedIntArray(indices),
+                size * SizeOf.SIZE_OF_INT);
+        input.readBytes(
+                Slices.wrappedDoubleArray(weights),
+                size * SizeOf.SIZE_OF_DOUBLE);
+        input.readBytes(
+                Slices.wrappedLongArray(counts),
+                size * SizeOf.SIZE_OF_LONG);
+    }
+
+    protected FixedDoubleBreakdownHistogram(FixedDoubleBreakdownHistogram other)
+    {
+        this.bucketCount = other.bucketCount;
+        this.min = other.min;
+        this.max = other.max;
+        Utils.verifyParams(bucketCount, min, max);
+        this.tolerance = other.tolerance;
+        size = other.size;
+        verifyParams();
+        this.indices = new int[size];
+        this.weights = new double[size];
+        this.counts = new long[size];
+        for (int i = 0; i < size; ++i) {
+            this.indices[i] = other.indices[i];
+            this.weights[i] = other.weights[i];
+            this.counts[i] = other.counts[i];
+        }
+    }
+
+    protected void verifyParams()
+    {
+        checkArgument(this.tolerance > 0, "Tolerance must be strictly positive");
```

Comment:
Does this need to be protected?  Also can we inline it?

Reply:
Made it private. I'd rather not inline it, as it is repeated a few times within the class.